calyx-opt/src/passes/well_formed.rs

```diff
@@ -600,14 +614,20 @@ impl Visitor for WellFormed {
         let cell = s.comp.borrow();
 
         if let CellType::Component { name: id } = &cell.prototype {
-            let cellmap = &self.ref_cell_types[id];
+            let cellmap = &self.ref_cells[id];
             let mut mentioned_cells = HashSet::new();
             for (outcell, incell) in s.ref_cells.iter() {
-                if let Some(t) = cellmap.get(outcell) {
-                    let proto = incell.borrow().prototype.clone();
-                    same_type(t, &proto)
-                        .map_err(|err| err.with_pos(&s.attributes))?;
-                    mentioned_cells.insert(outcell);
+                if let Some(oc) = cellmap.get(outcell) {
+                    if !subtype(oc, &incell.borrow()) {
+                        return Err(Error::malformed_control(format!(
+                            "The type passed in `{}` is not a subtype of the expected type `{}`.",
+                            incell.borrow().prototype.surface_name().unwrap(),
+                            oc.prototype.surface_name().unwrap()
+                        ))
+                        .with_pos(&s.attributes));
```

Comment:
Since this stanza appears twice, maybe we should outline it to a function `require_subtype` or something? Then the call here could just use `require_subtype(other)?` or similar to propagate an error.